src/rt/gtfsrt_update.cc

```diff
@@ -419,11 +432,135 @@ bool update_run(source_idx_t const src,
+      return geo::approx_squared_distance(stp.pos_, vp_position,
+                                          app_dist_lng_deg_vp) < 10;
+    });
+    if (stop_it == end(location_seq)) {
+      log(log_lvl::debug, "rt.gtfs.vehicle_update",
+          "Position of Vehicle was not near stop. Skipping message.", tag,
+          entity.id());
+      ++stats.vehicle_position_position_not_at_stop_;
+      return;
+    }
+
+    // get remaining stops
+    auto const stopped_at_idx =
+        stop_idx_t(std::distance(begin(location_seq), stop_it));
+    auto const fr = r.is_scheduled() ? frun::from_t(tt, &rtt_const, r.t_)
+                                     : frun::from_rt(tt, &rtt_const, r.rt_);
+    auto const seq_numbers = fr.stop_range_;
+
+    // get delay
+    auto const vp_ts = vp.has_timestamp()
+                           ? unixtime_t{std::chrono::duration_cast<i32_minutes>(
+                                 std::chrono::seconds{vp.timestamp()})}
+                           : std::chrono::system_clock::now();
+    auto const et = stopped_at_idx == 0
+                        ? tt.event_time(r.t_, stopped_at_idx, event_type::kDep)
+                        : tt.event_time(r.t_, stopped_at_idx, event_type::kArr);
+    auto const delay_cast = std::chrono::duration_cast<duration_t>(vp_ts - et);
+
+    // update delay for remaining stops
+    if (stopped_at_idx != *seq_numbers.begin()) {
+      update_delay(tt, rtt, r, stopped_at_idx, event_type::kArr, delay_cast,
+                   std::nullopt);
+    }
+    auto const stops_after = interval{stopped_at_idx, seq_numbers.to_};
+    for (auto const& [first, second] : utl::pairwise(stops_after)) {
+      update_delay(tt, rtt, r, first, event_type::kDep, delay_cast,
+                   std::nullopt);
+      update_delay(tt, rtt, r, second, event_type::kArr, delay_cast,
+                   std::nullopt);
+    }
+    if (stops_after.size() == 2) {
+      update_delay(tt, rtt, r, stopped_at_idx, event_type::kArr, delay_cast,
+                   std::nullopt);
+    }
+
+    // update delay for previous stops if necessary
+    auto const stops_before = interval{seq_numbers.from_, stopped_at_idx + 1};
+    if (stops_before.size() > 1) {
+      for (auto it = stops_before.rbegin();
+           it != std::prev(stops_before.rend()); ++it) {
+        auto prev = it;
+        ++prev;
+
+        if (rtt.unix_event_time(r.rt_, *prev, event_type::kDep) >
+            rtt.unix_event_time(r.rt_, *it, event_type::kArr)) {
+          update_delay(tt, rtt, r, *prev, event_type::kDep, delay_cast,
+                       std::nullopt);
+          if (prev != std::prev(stops_before.rend()) &&
+              rtt.unix_event_time(r.rt_, *prev, event_type::kArr) >
+                  rtt.unix_event_time(r.rt_, *prev, event_type::kDep)) {
+            update_delay(tt, rtt, r, *prev, event_type::kArr, delay_cast,
+                         std::nullopt);
+            continue;
+          }
+        }
+        break;
+      }
+    }
```

Comment:
this seems overly complex for what it does - can it be simplified for better readability and so it's easier to see if it's correct or not?

why not try something like pairwise(it_range{rbegin, rend})